### ISSUE TITLE
LoadBalancer Services cannot discover the true ClientIP

### ISSUE THREAD

**Issue by asauber** [linode/linode-cloud-controller-manager#74]
X-Forwarded-For, X-Real-IP, ProxyProtocol and friends currently see only an IP in the range 192.168.255.0/24

**Comment by asauber** [comment]
Yes, thank you for your comment. 😄 

ProxyProtocol support is coming soon to Linode NodeBalancers, which will allow all of these use cases either directly at the TCP level or beyond the Ingress controller. 🚀

**Comment by mortenbirkelund** [comment]
@asauber any news or time-line for when this will be supported? Thanks

**Comment by asauber** [comment]
Hi @mortenbirkelund! Proxy Protocol support was very recently added to NodeBalancers, so the next step is to add a Service annotation to this codebase that will allow enabling it. We don't have a specific timeline for this implementation at the moment. I will update this issue as soon as there is a PR or other progress on that.

**Comment by laszlocph** [comment]
:tada:

**Comment by asauber** [comment]
All LKE clusters will be upgraded to use this version of the CCM within the next two weeks.

**Comment by mortenbirkelund** [comment]
@asauber great news. Thanks for the update

**Comment by carminalazu** [comment]
Is there anything else that we need to do to benefit from this other than https://github.com/thechangelog/changelog.com/commit/a29bee7e47b7dd3aa5a0e66eb56dd14e310471e9 & then just waiting for this to be rolled out across Node Balancers?

**Comment by asauber** [comment]
@carminalazu In the manifest that you linked, you're placing the annotation on an Ingress resource. This annotation is used by a Service resource. In your case it would be the Service that sits in front of `ingress-nginx`. Once you've enabled Proxy Protocol for ingress-nginx, it looks like you would also need to set this setting in the ingress-nginx ConfigMap https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#use-proxy-protocol

**Comment by carminalazu** [comment]
I was assuming that this feature will make the `X-Forwarded-For`, `X-Real-IP` headers work as expected, which is what this captured:

https://github.com/thechangelog/changelog.com/blob/0e4e5831f494129d0697e34a54a5673e58c4491d/k8s/ingress-nginx/configmap.yml#L8-L12

The nginx-ingress docs explain that the Proxy protocol is a different approach:

> Another option is to enable proxy protocol using `use-proxy-protocol: "true"`.
> In this mode NGINX does not use the content of the header to get the source IP address of the connection.

This is the ingress-nginx config update that we should end up with:

```diff
-use-forwarded-headers: "true"
-
-# TODO: https://github.com/linode/linode-cloud-controller-manager/issues/74
+use-proxy-protocol: "true"
enable-real-ip: "true"
proxy-real-ip-cidr: "192.168.255.0/24"
```

If I understand you correctly, the annotation needs to be put on this service. Is that right?

https://github.com/kubernetes/ingress-nginx/blob/274d09447e4735815846985c3133bbdc3f00a942/deploy/static/provider/cloud/deploy.yaml#L273-L301

One last question: what happens if the PROXY protocol is enabled on the nginx ingress, but the node balancers don't have this capability rolled out?

**Comment by asauber** [comment]
This feature adds an annotation for LoadBalancer Services which enables Proxy Protocol on the associated NodeBalancers. This means that an Ingress controller that supports Proxy Protocol, like nginx-ingress, can then provide X-Forwarded-For headers to its backends.

Yes, you would need to modify that Service resource to add the annotation `service.beta.kubernetes.io/linode-loadbalancer-proxy-protocol: "v2"` Also, your diff looks correct, which disables header forwarding and enables Proxy Protocol.

To answer your last question. I'm happy to report that **as of October 2nd all Linode NodeBalancers and LKE clusters support these features!**

**Comment by asauber** [comment]
A description for this Proxy Protocol annotation can be found in the top level Readme https://github.com/linode/linode-cloud-controller-manager#annotations

**Comment by mortenbirkelund** [comment]
@asauber After enabling proxy, I am having some problems. I have a registy hosted on the same kubernetes cluster, and after enabling proxy protocol, it is no longer possible to pull images from the registry. 

When looking into the Ingress-controller pod I see the following errors when trying to pull an image. 

I have anonymized the domain and the ip
```
2020/10/14 09:41:40 [error] 7943#7943: *27088288 broken header: "�������P+"/J	i]Kp{=�-�(�T��b۬6�,�0�+�/�
�	��Xmy.registy.domain.com

" while reading PROXY protocol, client: 11.1.111.11, server: 0.0.0.0:443
2020/10/14 09:41:53 [error] 7943#7943: *27088369 broken header: "����t�yfh�!1ϓI�W6�g����o�֏�,�0�+�/�
�	��Xmy.registy.domain.com

" while reading PROXY protocol, client: 11.1.111.11, server: 0.0.0.0:443
2020/10/14 09:42:18 [error] 7943#7943: *27088525 broken header: "����g���5ِf�}EmY/�x2VO�^��,�0�+�/�
�	��Xmy.registy.domain.com

" while reading PROXY protocol, client: 11.1.111.11, server: 0.0.0.0:443
2020/10/14 09:43:02 [error] 7943#7943: *27088799 broken header: "���<��qʃ�����Q0��eX��O�zt���X�,�0�+�/�
�	��Xmy.registy.domain.com

" while reading PROXY protocol, client: 11.1.111.11, server: 0.0.0.0:443
2020/10/14 09:44:36 [error] 7942#7942: *27089382 broken header: "��y�&���{��p��Z���0MxɠS��H�h�D�,�0�+�/�
�	��Xmy.registy.domain.com
```

As soon as I disable proxy protocol in the ingress controller, everything works fine.

**Comment by laszlocph** [comment]
> @asauber After enabling proxy, I am having some problems. I have a registy hosted on the same kubernetes cluster, and after enabling proxy protocol, it is no longer possible to pull images from the registry.

The ingress controller must know about proxy protocol too. For Nginx it's this flag: https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#use-proxy-protocol

**Comment by mortenbirkelund** [comment]
@laszlocph Thank you for the quick reply. I have allready added this to the configmap of my ingress controller. 

Unfortunately it looks like it is actually a problem with Kubernetes and how it deals with external loadbalancers. 
Here is an explanation from digital ocean along with their workaround. Do you think it could be related?

https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/master/docs/controllers/services/examples/README.md#accessing-pods-over-a-managed-load-balancer-from-inside-the-cluster

https://github.com/kubernetes/kubernetes/issues/66607#issuecomment-474513060

**Comment by laszlocph** [comment]
> Here is an explanation from digital ocean along with their workaround. Do you think it could be related?

100%. I was facing the same issue on DO and `service.beta.kubernetes.io/do-loadbalancer-hostname: "{{ .Nginx.Host }}"` solved it. I'm about to try the Proxy Protocol on Linode and likely will hit the same issue. I think Linode has to solve it too somehow.

**Comment by mortenbirkelund** [comment]
I could imagine that having to wait for linode to implement a workaround or the kubernetes project to deal with this, will take some time. I guess i will have to be patient, or consider moving my registry to a separate cluser behind a separate loadbalancer.

**Comment by mortenbirkelund** [comment]
@asauber Should I create a new separate issue, or should this one be repopened as it is somewhat related to the newly implemented proxy protocol?

**Comment by asauber** [comment]
For a complete walkthrough of using Proxy Protocol with LoadBalancer services on LKE, check out this tutorial:
https://github.com/asauber/kubernetes-proxy-protocol-demo

Your in-cluster service which is using the container registry, is attempting to reach that Service via the hostname specified in the Ingress. 

Instead, it should use the in-cluster DNS name for the registry Service. This bypasses the Ingress controller, and is the preferred method for Services inside the cluster to reach each other. For example, with a minimal demo service running `httpbin`, this is what it looks like when we try to reach the service using the Ingress hostname.

```
$ kubectl run myshell --image busybox --command sleep 10000
$ kubectl exec -ti myshell -- sh
/ # wget http://mybackend.mydomain.com/get
Connecting to mybackend.mydomain.com (<public-ip>:80)
wget: error getting response: Resource temporarily unavailable
/ # exit
$ kubectl logs ingress-nginx-1602681250-controller-8df8684fc-5xbmd
...
2020/10/14 14:14:37 [error] 187#187: *11787 broken header:
...
```

These `broken header:` messages indicate that the in-cluster client is not sending Proxy Protocol data to ingress-nginx, which now expects it.

This is what we see when we use the in-cluster Service hostname

```
$ kubectl run myshell --image busybox --command sleep 10000
$ kubectl exec -ti myshell -- sh
/ # wget http://mybackend-service.default.svc.cluster.local/get
/ # cat get
{
  "args": {},
  "headers": {
    "Connection": "close",
    "Host": "mybackend-service.default.svc.cluster.local",
    "User-Agent": "Wget"
  },
  "origin": "10.2.1.9",
  "url": "http://mybackend-service.default.svc.cluster.local/get"
}
```

You will see that in this case the request succeeded and that the client IP is the Pod IP (of this busybox Pod) inside the cluster.

Here `svc.cluster.local` is the default domain name for the Service network in Kubernetes, you should be able to use that portion of the hostname without any modifications. Similarly, `default` is the name of the default namespace in Kubernetes, if your backend resides in a different namespace, then you can substitute that namespace name in this URL. The DNS records for this hostname are queried automatically via CoreDNS running in this cluster.

At this point you are able to reach the backend from both public and private clients:

- Publicly via http://mybackend.mydomain.com/
- Inside the cluster via http://mybackend-service.mybackend-namespace.svc.cluster.local/

Service hostnames should always be the preferred way to reach services inside a Kubernetes cluster. This is the mechanism for service discovery in Kubernetes.

**Comment by laszlocph** [comment]
@abemassry Thank you for your detailed comment. That behavior is a well documented k8s best practice, and the developer can decide whether she wants to go through the ingress using the external URL, or reaching the service on the internal service domain name.

With the container registry it can be circumvented with the internal domain, but it is only a workaround and many discussions in dev teams could be spared if Kubernetes wouldn't break such a fundamental thing in the name of optimization...

There are cases though when staying within the cluster is not possible or wanted: Cert Manager for example does not work on LKS as of today if Proxy Protocol is enabled. It's the most common way of obtaining Let's Encrypt certificates on Kubernetes and it works in a way that it tries to validate the external domain name from within the cluster, hitting the ingress controller directly but the ingress controller expects a proxy protocol input, thus the Let's Encrypt certs are never provisioned.

This deficiency in upstream k8s is solved in many clouds today either transparently, or by adding an annotation or alike. While upstream k8s is working on the case, I believe it would be great if a prevention of this routing bug would be solved on Linode :v:

**Comment by mortenbirkelund** [comment]
@asauber thanks for the good explanation 

It seems like my docker registry doesn't allow setting multiple hostname (I am using harbor) https://github.com/goharbor/harbor/issues/8243

So I guess I will either have to move my registry to a separate cluster or hope that you will implement a fix as suggested by @laszlocph

**Comment by jittuu** [comment]
I am now facing the cert-manager issue. The challenge is failed if Proxy Protocol is enabled.

I guess I have to wait for a while to use proxy protocol.

**Comment by asauber** [comment]
Hi. A quick update for this issue. We have planned to implement the hostname-based fix for this within the coming weeks.

In the mean time, users have reported success using this controller-based solution https://github.com/compumike/hairpin-proxy

**Comment by mortenbirkelund** [comment]
@asauber any idea when this will be implemented. I see there is an open pull request for it.

**Comment by sibucan** [comment]
@mortenbirkelund Hi there! Once the the PR is merged and a new CCM release is made with those changes, an LKE update will follow that will include an update to the CCM to the latest released version. We will ensure that when the update is available, the LKE changelog will reflect the CCM version being used. Thanks!

**Comment by mortenbirkelund** [comment]
Tried to use the newly added annotation `service.beta.kubernetes.io/linode-loadbalancer-hostname-only-ingress: "true"` but it doesnt seem to work. cert-manager is still failing because of self-check not going through. 

https://github.com/compumike/hairpin-proxy however did solve the problem.

**Comment by calbot** [comment]
I'm still very confused. Is this truly fixed? 
Is there some documentation for the correct encantations to get X-Forwarded-For to send the client IP when using helm to install the nginx ingress controller?

`helm upgrade --install ingress-nginx -f helm/inginx-ingress/customvalues.yaml ingress-nginx/ingress-nginx`

my customvalues.yaml file:
```
controller:
  service:
    annotations:
      service.beta.kubernetes.io/do-loadbalancer-enable-proxy-protocol: "true"
  config:
    use-forwarded-headers: "true"
    enable-real-ip: "true"
    proxy-real-ip-cidr: "192.168.255.0/24"
```

Can the documentation here  https://www.linode.com/docs/guides/how-to-deploy-nginx-ingress-on-linode-kubernetes-engine/
be updated so that X-Forwarded-For is properly set?

**Comment by calbot** [comment]
I was able to figure it out. If anyone else is using the helm chart instructions (from https://www.linode.com/docs/guides/how-to-deploy-nginx-ingress-on-linode-kubernetes-engine/) you can do the following... 
Log into the linode console > Nodebalancers > instanceID > configuration > 443/80 > Port Configuration > Proxy Protocol. I selected v2

Installing the helm chart below automatically creates the node balancer but doesn't setup the proxy protocol as above (for me at least).

The following configuration for the helm chart worked for me:
`helm upgrade --install ingress-nginx -f helm/inginx-ingress/customvalues.yaml ingress-nginx/ingress-nginx`

customvalues.yaml
```
controller:
  service:
    annotations:
      service.beta.kubernetes.io/do-loadbalancer-enable-proxy-protocol: "true"
      service.beta.kubernetes.io/linode-loadbalancer-proxy-protocol: "v2"
  config:
    use-proxy-protocol: "true"    
    use-forwarded-headers: "true"
```

After that my upstream apps received the correct `X-Forwarded-For`

**Comment by badrdouah** [comment]
> service.beta.kubernetes.io/do-loadbalancer-enable-proxy-protocol: "true"

i think you mean 

>  service.beta.kubernetes.io/linode-loadbalancer-enable-proxy-protocol: "true"


"do" stands for digital ocean,  this line in not nessecary, the second one will do the job i guess

**Comment by LukasBlu1** [comment]
Hi all! Is this also working with nodejs?

**Comment by shinebayar-g** [comment]
I'm running into this. Enabling proxy-protocol works. But cert-manager http01 validation breaks. It looks like this can be fixed with `hostname-only-ingress: 'true'` annotation. If I do this does it create a new nodebalancer? What will happen to the existing IP address? If IP address is lost we cannot enable this option as there are few customers that are using the NodeBalancer's IP address directly.

Or is there a way to get the IP address without proxy protocol?
I tried `externalTrafficPolicy: Local` and following configmap with the ingress-nginx. But it's still not working :(

```yaml
use-forwarded-headers: 'true'
enable-real-ip: 'true'
proxy-real-ip-cidr: '192.168.255.0/24'
```
